### PR TITLE
[Mono.Android, Xamarin.Android.Net.AndroidClientHandler] Fix requests with content

### PR DESCRIPTION
**Pull request by blounty**
Updated how content is handled, it seems content was not previously being added.

There also appeared to be a race condition which sometimes caused the connection output stream to no be ready at the time of writing.

### CODE REVIEW

File: src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs [hunk 1]
```diff
@@ -475,110 +474,126 @@ HttpURLConnection SetupRequestInternal (HttpRequestMessage request, URLConnectio
...
+                        return;
+
+                    tmf = TrustManagerFactory.GetInstance (TrustManagerFactory.DefaultAlgorithm);
+                    tmf.Init (keyStore);
+                }
+
+                SSLContext context = SSLContext.GetInstance ("TLS");
+                context.Init (kmf?.GetKeyManagers (), tmf.GetTrustManagers (), null);
+                httpsConnection.SSLSocketFactory = context.SocketFactory;
+            });
```
Comment (jonpryor) [inline src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs]: Tabs. vs. spaces mismatch. Please use tabs.

**Comment by jonpryor** [comment]
Tests for this should be added to <https://github.com/xamarin/xamarin-android/blob/master/src/Mono.Android/Test/Xamarin.Android.Net/AndroidClientHandlerTests.cs>.

File: src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs [hunk 2]
```diff
@@ -475,110 +474,126 @@ HttpURLConnection SetupRequestInternal (HttpRequestMessage request, URLConnectio
...
+                if (authorization == null) {
+                   if (Logger.Log)
+                        Logger.Log (LogLevel.Info, LOG_APP,$"Authorization module {auth.GetType ()} for scheme {data.Scheme} returned no authorization");
+                    return;
+                }
+
+               if (Logger.Log)
+                    Logger.Log (LogLevel.Info, LOG_APP,$"Authentication header '{data.UseProxyAuthentication ? "Proxy-Authorization" : "Authorization"}' will be set to '{authorization.Message}'");
+                httpConnection.SetRequestProperty (data.UseProxyAuthentication ? "Proxy-Authorization" : "Authorization", authorization.Message);
+            });
```
Comment (grendello) [inline src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs]: Tabs vs spaces here as well

**Comment by grendello** [comment]
LGTM - after tabs vs spaces are fixed and tests added

File: src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs [hunk 3]
```diff
@@ -516,69 +514,89 @@ void SetupSSL (HttpsURLConnection httpsConnection)
...
 		{
+			return Task.Factory.StartNew (() => {
 			AuthenticationData data = PreAuthenticationData;
 			if (!PreAuthenticate || data == null)
 				return;
 
 			ICredentials creds = data.UseProxyAuthentication ? Proxy?.Credentials : Credentials;
 			if (creds == null) {
-				if (Logger.LogNet)
+					if (Logger.LogNet)
```
Comment (jonpryor) [inline src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs]: Why is this indented?

File: src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs [hunk 4]
```diff
@@ -516,69 +514,89 @@ void SetupSSL (HttpsURLConnection httpsConnection)
...
 					Logger.Log (LogLevel.Info, LOG_APP, $"Authentication using scheme {data.Scheme} requested but no credentials found. No authentication will be performed");
 				return;
 			}
 
 			IAndroidAuthenticationModule auth = data.Scheme == AuthenticationScheme.Unsupported ? data.AuthModule : authModules.Find (m => m?.Scheme == data.Scheme);
 			if (auth == null) {
-				if (Logger.LogNet)
+					if (Logger.LogNet)
```
Comment (jonpryor) [inline src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs]: This shouldn't be indented this much.

File: src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs [hunk 5]
```diff
@@ -516,69 +514,89 @@ void SetupSSL (HttpsURLConnection httpsConnection)
...
 					Logger.Log (LogLevel.Info, LOG_APP, $"Authentication module for scheme '{data.Scheme}' not found. No authentication will be performed");
 				return;
 			}
 
 			Authorization authorization = auth.Authenticate (data.Challenge, httpConnection, creds);
 			if (authorization == null) {
-				if (Logger.LogNet)
+					if (Logger.LogNet)
```
Comment (jonpryor) [inline src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs]: Ditto.

File: src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs [hunk 6]
```diff
@@ -516,69 +514,89 @@ void SetupSSL (HttpsURLConnection httpsConnection)
...
 					Logger.Log (LogLevel.Info, LOG_APP, $"Authorization module {auth.GetType ()} for scheme {data.Scheme} returned no authorization");
 				return;
 			}
 
-			if (Logger.LogNet)
+				if (Logger.LogNet)
```
Comment (jonpryor) [inline src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs]: Ditto.

File: src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs [hunk 7]
```diff
@@ -516,69 +514,89 @@ void SetupSSL (HttpsURLConnection httpsConnection)
...
 			
-			conn.DoOutput = true;
-			long? contentLength = request.Content.Headers.ContentLength;
-			if (contentLength != null)
-				conn.SetFixedLengthStreamingMode ((int)contentLength);
+			httpConnection.DoOutput = true;
+
+			var bytes = await request.Content.ReadAsByteArrayAsync ().ConfigureAwait (false);
+
+			int contentLength = bytes.Count ();
```
Comment (jonpryor) [inline src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs]: Why is this `bytes.Count()` and not `bytes.Length`?

File: src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs [hunk 8]
```diff
@@ -516,69 +514,89 @@ void SetupSSL (HttpsURLConnection httpsConnection)
...
+			httpConnection.SetRequestProperty ("Content-Length", $"{contentLength}");
```
Comment (jonpryor) [inline src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs]: Why not `contentLength.ToString()`?

File: src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs (hunk 7 above)
Comment (blounty) [inline src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs]: No reason, would you prefer Length?

File: src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs (hunk 8 above)
Comment (blounty) [inline src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs]: No reason, i just like the style. Happy to change if you prefer ToString()

File: src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs (hunk 7 above)
Comment (jonpryor) [inline src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs]: Yes, I'd prefer `.Length`.

**Comment by jonpryor** [comment]
This still needs unit tests. :-)

**Comment by grendello** [comment]
@blounty I'm going to accept this seeing that more and more people run into this issue, however please provide the tests whenever you have the chance, thanks :)

**Comment by danielcaceresm** [comment]
I've solved this issue in my fork: https://github.com/danielcaceresm/xamarin-android/commit/bd1f0b4a40e65039b5330e24440c0cd46a5c1032
These changes are based on the AndroidClientHandler class before the modifications of this pull request.
I think that my changes are simplier, and the commented race condition, doesn't exists.
Other change I've made is to move the connection reference to the response class, this approach is better than holding the last connection in the handler. On the other hand, is pending to implement, dispose the connection if it fails while connecting.
If you consider my changes, I can open a new pull request.

**Comment by grendello** [comment]
@danielcaceresm please open a separete PR

**Pull request merged**